File: 8/8a.py
with open("8.txt") as file:
    rows = file.read().split('\n')

    res = 0

    visible = dict()

    for i in range(1, len(rows)-1):
        maxLtoR = int(rows[i][0])
        for j in range(1, len(rows[0])-1):
            height = int(rows[i][j])
            if height > maxLtoR:
                maxLtoR = height
                visible[f'{i},{j}'] = True
                res += 1

        maxRtoL = int(rows[i][-1])
        for j in range(len(rows[0])-2, 0, -1):
            height = int(rows[i][j])
            if height > maxRtoL:
                maxRtoL = height
                if f'{i},{j}' not in visible:
                    visible[f'{i},{j}'] = True
                    res += 1

    for j in range(1, len(rows[0])-1):
        maxTtoB = int(rows[0][j])
        for i in range(1, len(rows)-1):
            height = int(rows[i][j])
            if height > maxTtoB:
                maxTtoB = height
                if f'{i},{j}' not in visible:
                    visible[f'{i},{j}'] = True
                    res += 1

        maxBtoT = int(rows[-1][j])
        for i in range(len(rows)-2, 0, -1):
            height = int(rows[i][j])
            if height > maxBtoT:
                maxBtoT = height
                if f'{i},{j}' not in visible:
                    visible[f'{i},{j}'] = True
                    res += 1

    res += 2*len(rows[0])+2*(len(rows)-2)

    print(res)

File: 8/8a.js
const fs = require('fs');

const input = fs.readFileSync('./8.txt', 'utf8');

const rows = input.split(/\r?\n/);

let res = 0;

const visible = {};

// check for visible trees left to right and right to left
for (let i = 1; i < rows.length - 1; i++) {
  let maxLtoR = parseInt(rows[i][0], 10);
  for (let j = 1; j < rows[0].length - 1; j++) {
    let height = parseInt(rows[i][j], 10);
    if (height > maxLtoR) {
      maxLtoR = height;
      visible[i + ',' + j] = true;
      res += 1;
    }
  }

  let maxRtoL = parseInt(rows[i][rows[0].length - 1], 10);
  for (let j = rows[0].length - 2; j > 0; j--) {
    let height = parseInt(rows[i][j], 10);
    if (height > maxRtoL) {
      maxRtoL = height;
      // check if tree was counted as visible already
      if (!visible[i + ',' + j]) {
        visible[i + ',' + j] = true;
        res += 1;
      }
    }
  }
}

// check for visible trees top to bottom and bottom to top
for (let j = 1; j < rows[0].length - 1; j++) {
  let maxTtoB = parseInt(rows[0][j], 10);
  for (let i = 1; i < rows.length - 1; i++) {
    let height = parseInt(rows[i][j], 10);
    if (height > maxTtoB) {
      maxTtoB = height;
      // check if tree was counted as visible already
      if (!visible[i + ',' + j]) {
        visible[i + ',' + j] = true;
        res += 1;
      }
    }
  }

  let maxBtoT = parseInt(rows[rows.length - 1][j], 10);
  for (let i = rows.length - 2; i > 0; i--) {
    let height = parseInt(rows[i][j], 10);
    if (height > maxBtoT) {
      maxBtoT = height;
      // check if tree was counted as visible already
      if (!visible[i + ',' + j]) {
        visible[i + ',' + j] = true;
        res += 1;
      }
    }
  }
}

// add the borders to the final result
res += 2 * rows[0].length + 2 * (rows.length - 2);

console.log(res);
